### PR TITLE
feat: upload multiple asset chunks in the same call

### PR DESCRIPTION
**Pull request by sesi200**
# Description

The frontend canister sync now tries to batch multiple small content chunks into a single call using the `create_chunks` method added in #3898.
This should lead to significantly faster upload times for frontends with many small files.

Also had to bump the `api_version` of the asset canister to build in backwards compatibility for the asset sync

Fixes [SDK-1769](https://dfinity.atlassian.net/browse/SDK-1769)

# How Has This Been Tested?

Covered by existing e2e.
Manually tested repeatedly with many different average file sizes. Command to test:
```
for i in $(seq 1 50); do dd if=/dev/urandom of="src/hello_frontend/assets/file_$i.bin" bs=$(shuf -i 1-2000000 -n 1) count=1; done && dfx deploy hello_frontend
```

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

[SDK-1769]: https://dfinity.atlassian.net/browse/SDK-1769?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by chenyan-dfinity** [comment]
Is `icx-asset` updated as well? I tested this with one of the caffeine generated directory: https://github.com/chenyan-dfinity/asset-test/tree/main/1723158607637/frontend/dist

```
$ time ~/src/sdk/target/release/icx-asset upload br5f7-7uaaa-aaaaa-qaaca-cai dist/
```

On master, it takes 3.475s. On this branch, it takes 3.833s.

My simple chunk upload binary takes 2.158s. I expect icx-asset to be a bit slower because I didn't compute hash and didn't provide gzip files. But at least, it should be faster than master?

**Comment by sesi200** [comment]
> Is `icx-asset` updated as well?

Yes, it's 'just' a wrapper around `ic-asset`.

Just tested myself. Local with `dfx start --clean --artificial-delay 500` since I ran out of ICP to test on mainnet. Should be simulating a subnet with no load and 2 blocks per second, which should be close-ish to current ideal mainnet performance.

```
Project: dfx new, react frontend
icx-asset-pr:       5.38s
icx-asset-master:   5.35s

Project: dfx new, react frontend, plus 500 random files between 0 and 20k bytes, same files for both runs
icx-asset-pr:       8.45s
icx-asset-master:  34.14s
```

After verifying with debug prints in the asset canister I am pretty certain this happens:

`icx-asset-master`, hello world:
1. round: `create_batch`
2. round: 5 parallel calls to `create_chunk`, all executed in the same round (debug prints show same timestamp)
3. round: `commit_batch`

`icx-asset-pr`, hello world:
1. round: `create_batch`
2. round: `create_chunks` with 5 chunks in it
3. round: `commit_batch`

`icx-asset-master`, hello world + 500 files:
* 1. round: `create_batch`
* then: 500 times `create_chunk`, processes ~8 calls per round
* last 3 rounds: `commit_batch` since dfx only commits so many changes at a time (execution limits), all wait for the previous call's response

`icx-asset-pr`, hello world + 500 files:
* 1. round: `create_batch`
* 2. round: three calls to `create_chunks` processed in parallel, all 500 files uploaded
* 3-5. round: `commit_batch` since dfx only commits so many changes at a time, all wait for the previous call's response

So long story short: it speeds things up significantly, but only with enough files

File: src/canisters/frontend/ic-asset/src/batch_upload/plumbing.rs [hunk 1]
```diff
@@ -46,30 +49,69 @@ pub(crate) struct ProjectAsset {
...
+    ) -> Result<usize, CreateChunkError> {
+        let uploader_chunk_id = self.chunks.fetch_add(1, Ordering::SeqCst);
         self.bytes.fetch_add(contents.len(), Ordering::SeqCst);
-        create_chunk(&self.canister, &self.batch_id, contents, semaphores).await
+        if contents.len() == MAX_CHUNK_SIZE || self.api_version < 2 {
+            let canister_chunk_id =
+                create_chunk(&self.canister, &self.batch_id, contents, semaphores).await?;
+            let mut map = self.id_mapping.lock().await;
+            map.insert(uploader_chunk_id, canister_chunk_id);
+            println!("directly uploaded {uploader_chunk_id}");
```
Comment (adamspofford-dfinity) [inline src/canisters/frontend/ic-asset/src/batch_upload/plumbing.rs:93]: intentionally left in? If so, this should probably be `info!`

Comment (sesi200) [reply src/canisters/frontend/ic-asset/src/batch_upload/plumbing.rs:93]: That's an oversight, thanks!

**Review by adamspofford-dfinity — APPROVED** [review]
